### ISSUE TITLE
Bug mixing DMatrix's with and without feature names

### ISSUE THREAD

**Issue by flacjacket** [dmlc/xgboost#1101]
When a Booster is trained with a DMatrix with the default labels, the `.predict()` function does not throw an exception if you try to use the booster to perform prediction on a labeled DMatrix, and furthermore, after you have done that, you can no longer use the booster to predict the un-labeled DMatrix anymore, as if using the labeled DMatrix has changed the feature names the booster is looking for when it goes to predict.

Here is a minimal example:
```
import numpy as np
import xgboost as xgb

X = np.random.random((10, 3))
y = np.random.randint(2, size=(10,))

dm1 = xgb.DMatrix(X, y)
dm2 = xgb.DMatrix(X, y, feature_names=("a", "b", "c"))

bst = xgb.train([], dm1)
bst.predict(dm1)  # this works as expected
bst.predict(dm2)  # this works, but probably shouldn't
bst.predict(dm1)  # this now throws a ValueError
```

This seems like it could mainly be an issue when using the sklearn wrapper, like `XGBClassifier`, since the user does not control the initialization of the DMatrix objects, but by default, DMatrix's created from NumPy arrays will not have feature labels, and DMatrix's created from Pandas dataframes will have features from the column names.
